API Aggregator

chenopis opened the pull request:
[kubernetes/features#263](https://github.com/kubernetes/features/issues/263)

/assign @cheftako

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes.github.io/4173)
<!-- Reviewable:end -->

chenopis commented:
@lavalamp This is just the initial commit. I will fold in the changes from the Google doc once you guys have that ironed out.

chenopis commented:
@abiogenesis-now Can you help w/ a docs review? Thx

abiogenesis-now commented on docs/concepts/architecture/apiserver-aggregation.md:
```diff
@@ -0,0 +1,32 @@
+{% capture whatsnext %}
+
+* To get the aggregator working in your environment, see [Setup the API aggregator](/docs/tasks/access-kubernetes-api/setup-api-aggregator/).
+* Learn how to [Extend the Kubernetes API Using Third Party Resources](/docs/tasks/access-kubernetes-api/extend-api-third-party-resource/).
```
Looks like TPRs are being deprecated in 1.7? Swap out with [Extend the Kubernetes API Using Custom Resource Definitions](/docs/tasks/access-kubernetes-api/extend-api-custom-resource-definitions/).

abiogenesis-now replied:
Also, [this](https://docs.google.com/document/d/1lU1SnVtEec2iIfYx5U3L5N0za2YhfEOik0uPen276Ks/edit#heading=h.6ae9qgk4mhne) might be the Google doc that you are already aware of, but it seems like TPR/CRDs are distinct from API Aggregation? It would be great to make this more clear, since this is a high-level conceptual doc.

lavalamp replied:
Please don't call these third party resources. Extension APIs?

chenopis replied:
the link has been changed

abiogenesis-now commented on docs/concepts/architecture/apiserver-aggregation.md:
```diff
@@ -0,0 +1,32 @@
+
+{% capture overview %}
+
+The *Aggregator* is a system to allow the Kubernetes apiserver to be extended with additional APIs, which are not part of the core Kubernetes APIs. 
```
Is it best practice to refer to the "apiserver" or "API Server"? I've seen it both ways---it's more of a nit, but consistency would be nice.

abiogenesis-now replied:
I think the sentence reads a little easier as: "The *Aggregator* is a system that allows the Kubernetes apiserver to be extended with additional APIs, beyond what is offered by the core Kubernetes APIs."?

Let me know if this comment is too fine-grained.

lavalamp replied:
I would like to refer to this as the "aggregation layer" since it is built in and not a separate component.

lavalamp replied:
"The aggregation layer allows Kubernetes to be extended with additional APIs which are not included in the kubernetes core APIs"

chenopis replied:
Yeah, that makes a lot of sense.

abiogenesis-now commented on docs/tasks/access-kubernetes-api/setup-api-aggregator.md:
```diff
@@ -0,0 +1,66 @@
+
+{% capture whatsnext %}
+
+* For more information, see the [Aggregator Networking Requirements](https://docs.google.com/document/d/1KNT4iS_Y2miLARrfSPumBIiFo_h7eb5B2pVOZJ0ZmjQ/edit) doc.
```
I don't have access to this Google doc---might want to check the permissions on it.

chenopis replied:
I'm going to take this out until it's publicly accessible.

abiogenesis-now commented on docs/tasks/access-kubernetes-api/setup-api-aggregator.md:
```diff
@@ -0,0 +1,66 @@
+
+* For more information, see the [Aggregator Networking Requirements](https://docs.google.com/document/d/1KNT4iS_Y2miLARrfSPumBIiFo_h7eb5B2pVOZJ0ZmjQ/edit) doc.
+* For a high level overview, see [Extending the Kubernetes API with Aggregator](/docs/concepts/architecture/apiserver-aggregation/).
+* Learn how to [Extend the Kubernetes API Using Third Party Resources](/docs/tasks/access-Kubernetes-api/extend-api-third-party-resource/).
```
see the earlier comment about CRDs

chenopis replied:
got it

abiogenesis-now commented on docs/tasks/access-kubernetes-api/setup-api-aggregator.md:
```diff
@@ -0,0 +1,66 @@
+
+{% capture overview %}
+
+Setting up the API Aggregator will allow the Kubernetes apiserver to be extended with additional APIs, which are not part of the core Kubernetes APIs. 
```
nit: "allows" instead of "will allow" (there are a couple unnecessary future tenses scattered around...might want to just Ctrl-f "will")

lavalamp replied:
I would call this doc "turn on the aggregation layer" or "configure the aggregation layer" or "configure API Aggregation"

abiogenesis-now commented on docs/concepts/architecture/apiserver-aggregation.md:
```diff
@@ -0,0 +1,32 @@
+
+The Aggregator allows both third party extension via something like [service-catalog](https://github.com/kubernetes-incubator/service-catalog/blob/master/README.md) and the creation of extensions to allow seamless addition of the resources a customer needs. The latter can be easily bootstrapped using [apiserver-builder](https://github.com/kubernetes-incubator/apiserver-builder/blob/master/README. ...
+
+In 1.7, the Aggregator runs in a process with the kube-apiserver; however, until an extension resource is registered, the Aggregator will do nothing. Once a new resource is registered, Aggregator will proxy the request to the extension-apiserver to be served. 
```
nit: "to be served" might be redundant

abiogenesis-now replied:
also, is *extension-apiserver* an official term? I can infer what you're referring to, but it definitely made me pause. Would be nice to italicize it and mention it in the first paragraph.

cheftako replied:
Honestly I think we're still trying to come up with the official term. However I've heard several complaints about user-apiserver as that seems to exclude the third party use case. As this is meant as an extension mechanism we seem to be moving toward consensus on that term.

lavalamp replied:
My suggested wording in the doc was:

"In 1.7 the aggregation layer runs in-process with the kube-apiserver. Until an extension resource is registered, the aggregation layer will do nothing. Users can add a APIService object to register their API, which “claims” the URL path in the Kubernetes API. At this point, the aggregation layer will proxy anything sent to that api path (e.g. /apis/myextension.mycompany.io/v1/…) to the registered APIService. Ordinarily, the APIService will be implemented by an extension-apiserver in a pod running in the cluster. This extension-apiserver will normally need to be paired with one or more controllers if active management of the added resources is needed. As a result, the apiserver-builder will actually provide a skeleton for both. For example, when the service-catalog is installed, it provides both the extension-apiserver and controller for the services it provides."

chenopis replied:
swapped in

abiogenesis-now commented on docs/concepts/architecture/apiserver-aggregation.md:
```diff
@@ -0,0 +1,32 @@
+
+In 1.7, the Aggregator runs in a process with the kube-apiserver; however, until an extension resource is registered, the Aggregator will do nothing. Once a new resource is registered, Aggregator will proxy the request to the extension-apiserver to be served. 
+
+**Note:** The extension-apiserver will normally need to be paired with a controller to actually provide any functionality. As a result, things like the apiserver-builder will actually provide a skeleton for both. The service-catalog should provide both the extension-apiserver and controller for the services it provides.
```
^ a bit wordy----am I correctly interpreting that _both_ the apiserver-builder and the service-catalog provide the extension-apiserver and the controller? (A little confused if "should provide" means that service-catalog does it by default)

cheftako replied:
Yes

chenopis replied:
I've reworded this paragraph to clarify.

abiogenesis-now commented on docs/concepts/architecture/apiserver-aggregation.md:
```diff
@@ -0,0 +1,32 @@
+
+## Overview
+
+The Aggregator allows both third party extension via something like [service-catalog](https://github.com/kubernetes-incubator/service-catalog/blob/master/README.md) and the creation of extensions to allow seamless addition of the resources a customer needs. The latter can be easily bootstrapped using [apiserver-builder](https://github.com/kubernetes-incubator/apiserver-builder/blob/master/README. ...
```
It's not entirely clear to me what the difference between these two are, sorry! I'll read up on it later and see if I have suggestions, but do you mind clarifying?

cheftako replied:
Third party extensions are resource which are not built by either the cluster owner or the kubernetes team. Service Catalog is an example of this and is in fact an example which will make it easy for cluster owners to easily add other third party extensions to their cluster. This sort of extension is already built but does need to be deployed properly to the kubernetes cluster. If the cluster owner wants to create their own extension they first have to build something which adheres to the extension api-server requirements and then they have to work out how to deploy it correctly. So the story difference here is largely about the use case, though the technical details are very similar.

lavalamp replied:
"aggregation layer"

lavalamp replied:
"The aggregation layer allows installing additional Kubernetes-style APIs in your cluster. These can either be pre-built, existing 3rd party solutions (e.g., service catalog), or user-created APIs (apiserver-builder can get you started)."

abiogenesis-now replied:
yes this is much more clear!

chenopis replied:
changed

abiogenesis-now commented on docs/tasks/access-kubernetes-api/setup-api-aggregator.md:
```diff
@@ -0,0 +1,66 @@
+
+* You need to have a Kubernetes cluster running.
+* Make sure there are Aggregator certificates (both CA and proxy) setup on the Kubernetes apiserver.
+* The proxy cert must have been signed by the relevant CA. 
```
clarify what "relevant" means?

cheftako replied:
Kubernetes (and the kube-apiserver) have multiple CAs. I want to make sure the proxy is signed by the aggregator CA and not say the master CA.

abiogenesis-now replied:
Seems like this should go together with the previous bullet point then?

abiogenesis-now commented on docs/tasks/access-kubernetes-api/setup-api-aggregator.md:
```diff
@@ -0,0 +1,66 @@
+## Get an extension api-server working with Aggregator
+
+The generated skeleton, which comes with [apiserver-builder](https://github.com/Kubernetes-incubator/apiserver-builder/blob/master/README.md), should setup the skeleton server it creates to do all of these steps. You can also look at the sample-apiserver under Kubernetes for ways the following can be done.
+
```
provide link to sample-apiserver? Also consider rewording the first sentence.

cheftako replied:
https://github.com/kubernetes/sample-apiserver/blob/master/README.md

abiogenesis-now commented on docs/tasks/access-kubernetes-api/setup-api-aggregator.md:
```diff
@@ -0,0 +1,66 @@
+The generated skeleton, which comes with [apiserver-builder](https://github.com/Kubernetes-incubator/apiserver-builder/blob/master/README.md), should setup the skeleton server it creates to do all of these steps. You can also look at the sample-apiserver under Kubernetes for ways the following can be done.
+
+1. Create the Kubernetes namespace you want to run your extension api-service in.
+1. Create/get a CA cert to be used to sign the server cert the extension api-server uses for HTTPS.
```
do these easily map to code? sample commands might make it easier to follow.

chenopis commented:
@cheftako Can you suggest someone to do the tech review? Thx

cheftako commented:
I'm starting the review now.

On Fri, Jun 23, 2017 at 12:21 PM, Andrew Chen <notifications@github.com>
wrote:

> @cheftako <https://github.com/cheftako> Can you suggest someone to do the
> tech review? Thx
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes.github.io/pull/4173#issuecomment-310751826>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA53A_6BHWAtCfkWAsHjm0yMDxppTanhks5sHBAigaJpZM4OC-hM>
> .
>

cheftako commented on docs/tasks/access-kubernetes-api/setup-api-aggregator.md:
```diff
@@ -0,0 +1,68 @@
+    --proxy-client-cert-file=<path to aggregator proxy cert>
+    --proxy-client-key-file=<path to aggregator proxy key>
+
+If you are not running kube-proxy on your master, as the [Kubernetes Architectural Roadmap](https://docs.google.com/a/google.com/document/d/1XkjVm4bOeiVkj-Xt1LgoGiqWsBfNozJ51dyI-ljzt1o/edit?usp=sharing) recommends, then you need to make sure that the system is enabled with the following apiserver flag. Again, this may have already been taken care of by your provider.
```
I found this a tad confusing. Just to be clear the roadmap recommends not running kube-proxy on the master.

chenopis replied:
I reworded it

chenopis commented:
@cheftako Can we also get a another set of eyes on this for a second tech review?

cheftako commented:
@lavalamp has agreed to take a look.

On Fri, Jun 23, 2017 at 1:21 PM, Andrew Chen <notifications@github.com>
wrote:

> @cheftako <https://github.com/cheftako> Can we also get a another set of
> eyes on this for a second tech review?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes.github.io/pull/4173#issuecomment-310763872>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA53A6novLRrQSaXtJgssTooRyd7cN63ks5sHB5WgaJpZM4OC-hM>
> .
>

lavalamp commented on _data/concepts.yml:
```diff
@@ -20,6 +20,7 @@ toc:
   section:
   - docs/concepts/architecture/nodes.md
   - docs/concepts/architecture/master-node-communication.md
+  - docs/concepts/architecture/apiserver-aggregation.md
```
I agree this should probably go along with the api section, or we could make a new api extension section and include this, TPR, CRD, & admission controllers / initializers.

chenopis replied:
Ok, I'm going to create a new subsection under Concepts called *Extending the Kubernetes API* and put this in there. We can then start adding other docs into that. I did originally try to put it in the Overview subsection next to *Kubernetes API*, but this doc isn't really an overview so it didn't quite fit.

lavalamp commented on docs/tasks/access-kubernetes-api/setup-api-aggregator.md:
```diff
@@ -0,0 +1,68 @@
+
+{% endcapture %}
+
+{% capture prerequisites %}
```
These are not prereqs, these are what this doc is going to tell you how to do.

chenopis replied:
Ok. Are there any other prereqs then that they'll need before they can get started?

lavalamp commented on docs/tasks/access-kubernetes-api/setup-api-aggregator.md:
```diff
@@ -0,0 +1,68 @@
+
+    --enable-aggregator-routing=true
+
+## Get an extension api-server working with Aggregator
```
IMO this and everything below should go in a separate doc, "how to setup an extension API server".

chenopis replied:
Ok, I'm splitting this into two docs.

chenopis commented:
@lavalamp Ok, thanks for the feedback. I'll address your feedback this weekend.

chenopis commented:
I've addressed all the comments. @cheftako @lavalamp @abiogenesis-now PTAL.

Previews:
Concept/[Extending the Kubernetes API with the aggregation layer](https://deploy-preview-4173--kubernetes-io-master-staging.netlify.com/docs/concepts/api-extension/apiserver-aggregation/)
Task/[Configure the aggregation layer](https://deploy-preview-4173--kubernetes-io-master-staging.netlify.com/docs/tasks/access-kubernetes-api/configure-aggregation-layer/)
Task/[Setup an extension API server](https://deploy-preview-4173--kubernetes-io-master-staging.netlify.com/docs/tasks/access-kubernetes-api/setup-extension-api-server/)

chenopis commented:
FYI, all feedback must be addressed and LGTMs given by EOD Tue, June 27th so that this can be merged for the 1.7 release on June 28th.

abiogenesis-now commented on docs/concepts/api-extension/apiserver-aggregation.md:
```diff
@@ -0,0 +1,35 @@
+
+The aggregation layer enables installing additional Kubernetes-style APIs in your cluster. These can either be pre-built, existing 3rd party solutions, such as [service-catalog](https://github.com/kubernetes-incubator/service-catalog/blob/master/README.md), or user-created APIs like [apiserver-builder](https://github.com/kubernetes-incubator/apiserver-builder/blob/master/README.md), which can get ...
+
+In 1.7 the aggregation layer runs in-process with the kube-apiserver. Until an extension resource is registered, the aggregation layer will do nothing. Users can add an APIService object to register their API, which “claims” the URL path in the Kubernetes API. At that point, the aggregation layer will proxy anything sent to that API path (e.g. /apis/myextension.mycompany.io/v1/…) to the registere ...
```
want to rewrite the third sentence:

> To register their API, users must add an APIService object, which "claims" the URL path in the Kubernetes API.

in order to better contrast with the second sentence (the aggregation layer doing nothing unless extension resource is registered)

abiogenesis-now replied:
I want to gut check my own understanding of things, so quick question: 

**Is it the case that kube-apiserver is an actual official binary, whereas extension-apiserver is a stand-in name that can refer to *any* roll-your-own custom API server?**

(e.g. I should not expect to be able to literally run `./extension-apiserver` unless I explicitly name the custom API server binary that way?)

abiogenesis-now replied:
is the aggregation layer proxying mentioned here something that `kube-apiserver` does or `kube-proxy`?

chenopis replied:
@cheftako @lavalamp Can you clarify?

chenopis replied:
done

chenopis replied:
3rd sentence swapped out

abiogenesis-now commented on docs/tasks/access-kubernetes-api/configure-aggregation-layer.md:
```diff
@@ -0,0 +1,52 @@
+
+## Enable apiserver flags
+
+Enable the aggregation layer via the following apiserver flags. They may have already been taken care of by your provider.
```
*Clarification:* Are these flags meant to be applied when I run `kube-apiserver`?  

I'm pretty sure this is the case but I could be wrong. Either way, could we clarify that "apiserver" means `kube-apiserver`, given that there are now multiple API servers floating around?

chenopis replied:
@cheftako @lavalamp Can you confirm that "apiserver" means `kube-apiserver` here?

cheftako replied:
It does mean kube-apiserver.

chenopis replied:
Thanks for the clarification.

abiogenesis-now commented on docs/tasks/access-kubernetes-api/setup-extension-api-server.md:
```diff
@@ -0,0 +1,53 @@
+
+## Setup an extension api-server to work with the aggregation layer
+
+If you use an existing 3rd party solution, such as [apiserver-builder](https://github.com/Kubernetes-incubator/apiserver-builder/blob/master/README.md), it should generate a skeleton and perform all of the following steps. Alternatively, you can also look at the [sample-apiserver](https://github.com/kubernetes/sample-apiserver/blob/master/README.md) under Kubernetes for ways the following can be  ...
```
I was a bit confused by this blurb at first (e.g. why the steps below lacked detail). Consider the following wording for more clarity?

> The following steps describe how to set up an extension-apiserver *at a high level*. For a concrete example of how they can be implemented, you can look at the [sample-apiserver](https://github.com/kubernetes/sample-apiserver/blob/master/README.md) in the Kubernetes repo.
>
> Alternatively, you can use an existing 3rd party solution, such as [apiserver-builder](https://github.com/Kubernetes-incubator/apiserver-builder/blob/master/README.md), which should generate a skeleton and automate all of the following steps for you.

chenopis replied:
done

abiogenesis-now commented on docs/concepts/api-extension/apiserver-aggregation.md:
```diff
@@ -0,0 +1,35 @@
+
+In 1.7 the aggregation layer runs in-process with the kube-apiserver. Until an extension resource is registered, the aggregation layer will do nothing. Users can add an APIService object to register their API, which “claims” the URL path in the Kubernetes API. At that point, the aggregation layer will proxy anything sent to that API path (e.g. /apis/myextension.mycompany.io/v1/…) to the registere ...
+
+**Note:** The extension-apiserver will need to be paired with a controller to provide any functionality. Fortunately, if you use [apiserver-builder](https://github.com/kubernetes-incubator/apiserver-builder/blob/master/README.md), it will provide both a skeleton extension-apiserver and a skeleton controller. Alternatively, the [service-catalog](https://github.com/kubernetes-incubator/service-cata ...
```
by entire part I mean line 23 😅

chenopis replied:
done

chenopis commented on docs/tasks/access-kubernetes-api/configure-aggregation-layer.md:
```diff
@@ -0,0 +1,52 @@
+
+{% capture overview %}
+
+Configuring the aggregation layer allows the Kubernetes apiserver to be extended with additional APIs, which are not part of the core Kubernetes APIs. 
```
good idea

chenopis commented on docs/tasks/access-kubernetes-api/configure-aggregation-layer.md:
```diff
@@ -0,0 +1,52 @@
+{% capture whatsnext %}
+
+* [Setup an extension api-server](/docs/tasks/access-kubernetes-api/setup-extension-api-server/) to work with the aggregation layer.
+* For a high level overview, see [Extending the Kubernetes API with Aggregator](/docs/concepts/api-extension/apiserver-aggregation/).
```
Good catch. Sorry, was updating things out of order.

chenopis commented on docs/tasks/access-kubernetes-api/setup-extension-api-server.md:
```diff
@@ -0,0 +1,53 @@
+{% capture prerequisites %}
+
+* You need to have a Kubernetes cluster running.
+* You must [configure the aggregation layer](/docs/tasks/configure-aggregation-layer) and enable the apiserver flags.
```
good catch

chenopis commented on docs/tasks/access-kubernetes-api/setup-extension-api-server.md:
```diff
@@ -0,0 +1,53 @@
+
+If you use an existing 3rd party solution, such as [apiserver-builder](https://github.com/Kubernetes-incubator/apiserver-builder/blob/master/README.md), it should generate a skeleton and perform all of the following steps. Alternatively, you can also look at the [sample-apiserver](https://github.com/kubernetes/sample-apiserver/blob/master/README.md) under Kubernetes for ways the following can be  ...
+
+1. Make sure the APIService API is enabled (check --runtime-config); it is on by default, so it should be on unless it’s been deliberately turned off in your cluster.
```
done

chenopis commented on docs/tasks/access-kubernetes-api/setup-extension-api-server.md:
```diff
@@ -0,0 +1,53 @@
+{% capture whatsnext %}
+
+* If you haven't already, [configure the aggregation layer](/docs/tasks/access-kubernetes-api/configure-aggregation-layer/) and enable the apiserver flags.
+* For a high level overview, see [Extending the Kubernetes API with Aggregator](/docs/concepts/api-extension/apiserver-aggregation/).
```
duh. thanks!

chenopis commented:
@abiogenesis-now I think I put in all your feedback. Let me know if I missed anything.

chenopis commented:
@cheftako @lavalamp PTAL

abiogenesis-now commented:
LGTM---thanks for incorporating feedback!

cheftako commented on docs/concepts/api-extension/apiserver-aggregation.md:
```diff
@@ -0,0 +1,35 @@
+
+In 1.7 the aggregation layer runs in-process with the kube-apiserver. Until an extension resource is registered, the aggregation layer will do nothing. To register their API, users must add an APIService object, which "claims" the URL path in the Kubernetes API. At that point, the aggregation layer will proxy anything sent to that API path (e.g. /apis/myextension.mycompany.io/v1/…) to the registe ...
+
+Ordinarily, the APIService will be implemented by an *extension-apiserver* in a pod running in the cluster. This extension-apiserver will normally need to be paired with one or more controllers if active management of the added resources is needed. As a result, the apiserver-builder will actually provide a skeleton for both. For example, when the service-catalog is installed, it provides both the ...
```
As another example, when the service catalog

chenopis replied:
got it
